src/target/riscv/riscv-013.c

```diff
@@ -3146,10 +3130,33 @@ static int modify_privilege(struct target *target, uint64_t *mstatus, uint64_t *
 	*mstatus = set_field(*mstatus, MSTATUS_MPRV, 1);
 
 	/* Write MSTATUS */
-	if (*mstatus == *mstatus_old)
+	if (*mstatus != *mstatus_old &&
+			register_write_direct(target, GDB_REGNO_MSTATUS, *mstatus) != ERROR_OK)
+		return ERROR_FAIL;
+
+	/* dcsr.mprven <- 1 */
+	return register_write_direct(target, GDB_REGNO_DCSR,
+			set_field(dcsr, CSR_DCSR_MPRVEN, CSR_DCSR_MPRVEN_ENABLED));
+}
+
+static int restore_privilege(struct target *target, uint64_t mstatus, uint64_t mstatus_old)
+{
+	if (!riscv_virt2phys_mode_is_hw(target))
 		return ERROR_OK;
 
-	return register_write_direct(target, GDB_REGNO_MSTATUS, *mstatus);
+	/* Restore MSTATUS */
+	if (mstatus != mstatus_old &&
+		    register_write_direct(target, GDB_REGNO_MSTATUS, mstatus_old) != ERROR_OK)
+		return ERROR_FAIL;
+
+	/* Read DCSR */
+	riscv_reg_t dcsr;
+	if (register_read_direct(target, &dcsr, GDB_REGNO_DCSR) != ERROR_OK)
+		return ERROR_FAIL;
+
+	/* dcsr.mprven <- 0 */
```

Comment:
why do you write this to `zero unconditionally` ?  should not we preserve the value?

Reply:
`riscv_program_load_mprv()` set the `dcsr.mprven` to zero unconditionally, and I followed its logic. If I set `riscv virt2phys_mode` back to sw, `dcsr.mprven` should set to zero?

Reply:
> should not we preserve the original value?

Strictly speaking, I don't think we should -- `dcsr` value is managed by the debugger exclusively. Though either approach should work just fine.